Svc/BufferManager/BufferManagerComponentImpl.cpp

```diff
@@ -0,0 +1,222 @@
+              // change size to match request
+              copy.setSize(size);
+              return copy;
+          }
+      }
+
+      // if no buffers found, return empty buffer
+      this->log_WARNING_HI_NoBuffsAvailable(size);
+      this->m_noBuffs++;
+      return Fw::Buffer();
+
+  }
+
+  void BufferManagerComponentImpl::setup(
+    NATIVE_UINT_TYPE mgrId, //!< manager ID
+    NATIVE_UINT_TYPE memId, //!< Memory segment identifier
+    Fw::MemAllocator& allocator, //!< memory allocator
+    const BufferBins& bins //!< Set of user bins
+  ) {
+
+    this->m_mgrId = mgrId;
+    this->m_memId = memId;
+    this->m_allocator = &allocator;
+    // clear bins
+    memset(&this->m_bufferBins,0,sizeof(this->m_bufferBins));
+
+    this->m_bufferBins = bins;
+
+    // compute the amount of memory needed
+    NATIVE_UINT_TYPE memorySize = 0; // track needed memory
+    this->m_numStructs = 0; // size the number of tracking structs
+    // walk through bins and add up the sizes
+    for (NATIVE_UINT_TYPE bin = 0; bin < BUFFERMGR_MAX_NUM_BINS; bin++) {
+        if (this->m_bufferBins.bins[bin].numBuffers) {
+            memorySize += 
+                this->m_bufferBins.bins[bin].bufferSize * this->m_bufferBins.bins[bin].numBuffers // allocate each set of buffer memory
+                + sizeof(AllocatedBuffer) * this->m_bufferBins.bins[bin].numBuffers; // allocate the structs to track the buffers
+            this->m_numStructs += this->m_bufferBins.bins[bin].numBuffers;
+        }
+    }
+
+    NATIVE_UINT_TYPE allocatedSize = memorySize;
+    bool recoverable; //!< don't care if it is recoverable since they are a pool of user buffers
+
+    // allocate memory
+    void *memory = allocator.allocate(memId,allocatedSize,recoverable);
+    // make sure the memory returns was non-zero and the size requested
+    FW_ASSERT(memory);
+    FW_ASSERT(memorySize == allocatedSize,memorySize,allocatedSize);
+    // structs will be at beginning of memory
+    this->m_buffers = static_cast<AllocatedBuffer*>(memory);
+    // memory buffers will be at end of structs in memory, so compute that memory as the begining of the 
+    // struct past the number of structs
+    U8* bufferMem = reinterpret_cast<U8*>(&this->m_buffers[this->m_numStructs]);
+
+    // walk through entries and initialize them
+    NATIVE_UINT_TYPE currStruct = 0;
+    for (NATIVE_UINT_TYPE bin = 0; bin < BUFFERMGR_MAX_NUM_BINS; bin++) {
+        if (this->m_bufferBins.bins[bin].numBuffers) {
+            for (NATIVE_UINT_TYPE binEntry = 0; binEntry < this->m_bufferBins.bins[bin].numBuffers; binEntry++) {
+                // placement new for Fw::Buffer instance. We don't need the new() return value, 
+                // because we know where the Fw::Buffer instance is
+                U32 context = (this->m_mgrId << 16) | currStruct;
+                (void) new(&this->m_buffers[currStruct].buff) Fw::Buffer(bufferMem,this->m_bufferBins.bins[bin].bufferSize,context);
```

Comment:
This code is using more memory than it needs to. The `bufferMem` pointer is stored twice. The other information in the `Fw::Buffer` can be computed when the buffer is allocated. It doesn't need to be stored in the reserved memory area.

Reply:
Also, it's a good idea to assert that `currStruct` is in bounds before using it to access `m_buffers`.

Reply:
Will be followed under issue https://github.com/nasa/fprime/issues/340